### PR TITLE
[Dynamic Selection] auto_tune_policy

### PR DESCRIPTION
**Pull request by vossmjp**
Initial auto_tune implementation.

### CODE REVIEW

File: test/parallel_api/dynamic_selection/sycl/test_auto_tune_policy_sycl.pass.cpp [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+    std::cout << "select then submit_and_wait: OK\n";
+  } else {
+    std::cout << "submit_and_wait: OK\n";
+  }
+  return 0;
+}
+
+/*
+
+template<typename Policy, typename UniverseContainer, bool do_select=false>
```
Comment (AnuyaWelling2801) [inline test/parallel_api/dynamic_selection/sycl/test_auto_tune_policy_sycl.pass.cpp:340]: Delete commented out code

File: include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h
Comment (AnuyaWelling2801) [inline include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h:1]: The get_submission group should use the newer interface with waiters_

**Review by AnuyaWelling2801 — CHANGES_REQUESTED** [review]

File: include/oneapi/dpl/internal/dynamic_selection.h [hunk 2]
```diff
@@ -123,18 +130,41 @@ namespace experimental {
...
-        return std::forward<DSPolicy>(dp).submit_and_wait(e, std::forward<Function>(f), std::forward<Args>(args)...);
-    }else{
-        return wait(submit(std::forward<DSPolicy>(dp), e, std::forward<Function>(f), std::forward<Args>(args)...));
+  template<typename T, typename Function, typename... Args>
+  auto submit_and_wait(T&& t, Function&&f, Args&&... args) {
+    if constexpr (has_get_policy<T>::value == true) { 
+      // t is a selection
+      if constexpr (has_submit_and_wait_handle<decltype(std::declval<T>().get_policy()), T, Function, Args...>::value == true) {
+        // policy has optional submit_and_wait(selection, f, args...)
+        return t.get_policy().submit_and_wait(t, std::forward<Function>(f), std::forward<Args>(args)...);
```
Comment (egfefey) [inline include/oneapi/dpl/internal/dynamic_selection.h:139]: why not std::forward<T>(t)?

File: include/oneapi/dpl/internal/dynamic_selection.h [hunk 3]
```diff
@@ -123,18 +130,41 @@ namespace experimental {
...
+      } else {   
+        // policy does not have optional submit_and_wait for a selection
+        return wait(submit(t, std::forward<Function>(f), std::forward<Args>(args)...));
```
Comment (egfefey) [inline include/oneapi/dpl/internal/dynamic_selection.h:142]: why not std::forward(t)?

File: include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+#define _ONEDPL_AUTO_TUNE_POLICY_H
+
+#include <atomic>
+#include <exception>
+#include <map>
+#include <memory>
+#include <mutex>
+#include <utility>
+#include <vector>
+#include "oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h"
```
Comment (egfefey) [inline include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h:20]: What if I'm not using sycl?

File: include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+      }
+    }
+
+    auto_tune_policy(const std::vector<resource_type>& u, double resample_time=never_resample) {
+      if (resample_time != deferred_initialization) {
+        initialize(u, resample_time);
+      }
+    }
+   
+    void initialize(double resample_time=never_resample) {
```
Comment (egfefey) [inline include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h:157]: Spec calls the argument to initialize() offset, so maybe we change the spec to something generic to account for different uses in different policies

File: include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+      } 
+    }
+
+    template<typename Function, typename ...Args>
+    auto submit(selection_type e, Function&& f, Args&&... args) {
+      return backend_->submit(e, std::forward<Function>(f), std::forward<Args>(args)...);
+    }
+
+    auto get_resources() {
+       if (backend_) {
```
Comment (egfefey) [inline include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h:186]: Should we be checking for the size of the resources instead?

File: test/parallel_api/dynamic_selection/sycl/test_auto_tune_policy_sycl.pass.cpp (hunk 1 above)
Comment (vossmjp) [reply test/parallel_api/dynamic_selection/sycl/test_auto_tune_policy_sycl.pass.cpp:340]: Yep, meant to do that. Will do.

File: include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h
Comment (vossmjp) [reply include/oneapi/dpl/internal/dynamic_selection_impl/sycl_backend.h:1]: Let's do as part of another PR.

File: include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h (hunk 4 above)
Comment (vossmjp) [reply include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h:20]: the sycl_backend is the default. So this is necessary I think.

File: include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h (hunk 6 above)
Comment (vossmjp) [reply include/oneapi/dpl/internal/dynamic_selection_impl/auto_tune_policy.h:186]: At least in auto_tune, this is a test for deferred_initialization. No backend_ is created if deferred and not yet initialized.

File: include/oneapi/dpl/internal/dynamic_selection.h (hunk 3 above)
Comment (vossmjp) [reply include/oneapi/dpl/internal/dynamic_selection.h:142]: An oversight. Will fix.

**Review by egfefey — APPROVED** [review]

**Pull request merged**